### PR TITLE
Upgrade dependencies in `mock-block-dock` and `block-template`

### LINKED ISSUE #162: create-block-app could do with modernisation
The generated app feels a little outdated:
* It doesn't support nullish coalescing (in my case in an imported es6 third party module)
* Using old versions of webpack
* Uses outdated version of babel (why is it using babel in a typescript project?) 

FWIW I see the potential and have some components that would be very suitable with very little work, but the lack of nullish coalescing and my unfamiliarity with babel started to make it a chore.

I don't mean to be negative, and am sorry that the above is all I can contribute at the moment,

### PR DESCRIPTION
**Pull request by kachkaev**
This PR closes #162 by upgrading `block-template` dependencies. We now have webpack 5 in the toolchain, which provides a slightly better DX and opens opportunities for further improvements. Modern JS syntax like `??` seems to be working in the blocks – it compiles without errors.

---

We still get ESLint errors for the modern syntax, but that’s unrelated to the published blocks. Our repo still on `eslint` v7, which uses a pretty old version of `espree` parser. This will be addressed separately.

Screenshot 2022-02-09 at 00 04 43

---

There is another problem in packages, which is probably outside this PR scope. `yarn dev` for `mock-block-dock` is partially failing in both on `main` and in this branch.

<details>
<summary>Before (webpack 4)</summary>
<pre><code>yarn run v1.22.17
$ concurrently -n webpack,webpack-dev-server -c green,cyan "yarn build:dev -- --watch --verbose" "yarn run-dev-server"
warning From Yarn 1.0 onwards, scripts don't require "--" for options to be forwarded. In a future version, any explicit "--" will be forwarded as-is to the scripts.
$ cross-env NODE_ENV=development webpack-dev-server -d --port 9090 --config dev/webpack.config.js
$ clean; cross-env NODE_ENV=development webpack --mode development --watch --verbose
[webpack] /bin/sh: clean: command not found
[webpack] 
[webpack] webpack is watching the files...
[webpack] 
[webpack] 
[webpack] Insufficient number of arguments or no entry found.
[webpack] Alternatively, run 'webpack(-cli) --help' for usage info.
[webpack] 
[webpack] Hash: f51f0734ccb9579023e8
[webpack] Version: webpack 4.46.0
[webpack] Time: 16ms
[webpack] Built at: 02/09/2022 1:32:48 AM
[webpack] 
[webpack] LOG from webpack.buildChunkGraph.visitModules
[webpack] <t> prepare: 0.060083ms
[webpack] 
[webpack] ERROR in Entry module not found: Error: Can't resolve './src' in '/Users/ak/-/projects/hash/github/blockprotocol/packages/mock-block-dock'
[webpack] resolve './src' in '/Users/ak/-/projects/hash/github/blockprotocol/packages/mock-block-dock'
[webpack]   using description file: /Users/ak/-/projects/hash/github/blockprotocol/packages/mock-block-dock/package.json (relative path: .)
[webpack]     Field 'browser' doesn't contain a valid alias configuration
[webpack]     using description file: /Users/ak/-/projects/hash/github/blockprotocol/packages/mock-block-dock/package.json (relative path: ./src)
[webpack]       no extension
[webpack]         Field 'browser' doesn't contain a valid alias configuration
[webpack]         /Users/ak/-/projects/hash/github/blockprotocol/packages/mock-block-dock/src is not a file
[webpack]       .wasm
[webpack]         Field 'browser' doesn't contain a valid alias configuration
[webpack]         /Users/ak/-/projects/hash/github/blockprotocol/packages/mock-block-dock/src.wasm doesn't exist
[webpack]       .mjs
[webpack]         Field 'browser' doesn't contain a valid alias configuration
[webpack]         /Users/ak/-/projects/hash/github/blockprotocol/packages/mock-block-dock/src.mjs doesn't exist
[webpack]       .js
[webpack]         Field 'browser' doesn't contain a valid alias configuration
[webpack]         /Users/ak/-/projects/hash/github/blockprotocol/packages/mock-block-dock/src.js doesn't exist
[webpack]       .json
[webpack]         Field 'browser' doesn't contain a valid alias configuration
[webpack]         /Users/ak/-/projects/hash/github/blockprotocol/packages/mock-block-dock/src.json doesn't exist
[webpack]       as directory
[webpack]         existing directory
[webpack]           using path: /Users/ak/-/projects/hash/github/blockprotocol/packages/mock-block-dock/src/index
[webpack]             using description file: /Users/ak/-/projects/hash/github/blockprotocol/packages/mock-block-dock/package.json (relative path: ./src/index)
[webpack]               no extension
[webpack]                 Field 'browser' doesn't contain a valid alias configuration
[webpack]                 /Users/ak/-/projects/hash/github/blockprotocol/packages/mock-block-dock/src/index doesn't exist
[webpack]               .wasm
[webpack]                 Field 'browser' doesn't contain a valid alias configuration
[webpack]                 /Users/ak/-/projects/hash/github/blockprotocol/packages/mock-block-dock/src/index.wasm doesn't exist
[webpack]               .mjs
[webpack]                 Field 'browser' doesn't contain a valid alias configuration
[webpack]                 /Users/ak/-/projects/hash/github/blockprotocol/packages/mock-block-dock/src/index.mjs doesn't exist
[webpack]               .js
[webpack]                 Field 'browser' doesn't contain a valid alias configuration
[webpack]                 /Users/ak/-/projects/hash/github/blockprotocol/packages/mock-block-dock/src/index.js doesn't exist
[webpack]               .json
[webpack]                 Field 'browser' doesn't contain a valid alias configuration
[webpack]                 /Users/ak/-/projects/hash/github/blockprotocol/packages/mock-block-dock/src/index.json doesn't exist
[webpack-dev-server] ℹ ｢wds｣: Project is running at http://localhost:9090/
[webpack-dev-server] ℹ ｢wds｣: webpack output is served from /
[webpack-dev-server] ℹ ｢wds｣: Content not from webpack is served from /Users/ak/-/projects/hash/github/blockprotocol/packages/mock-block-dock/dev
[webpack-dev-server] ℹ ｢wdm｣: wait until bundle finished: /
[webpack-dev-server] ℹ ｢wdm｣: Hash: 4a3f4ccb866ed93a6104
[webpack-dev-server] Version: webpack 4.46.0
[webpack-dev-server] Time: 874ms
[webpack-dev-server] Built at: 02/09/2022 1:32:49 AM
[webpack-dev-server]      Asset       Size  Chunks             Chunk Names
[webpack-dev-server] index.html  225 bytes          [emitted]  
[webpack-dev-server]    main.js   3.86 MiB    main  [emitted]  main
[webpack-dev-server] Entrypoint main = main.js
[webpack-dev-server] [0] multi (webpack)-dev-server/client?http://localhost:9090 (webpack)/hot/dev-server.js ./dev/DevApp.tsx 52 bytes {main} [built]
[webpack-dev-server] [../../node_modules/react-dom/index.js] /Users/ak/-/projects/hash/github/blockprotocol/node_modules/react-dom/index.js 1.33 KiB {main} [built]
[webpack-dev-server] [../../node_modules/react/index.js] /Users/ak/-/projects/hash/github/blockprotocol/node_modules/react/index.js 190 bytes {main} [built]
[webpack-dev-server] [../../node_modules/webpack-dev-server/client/index.js?http://localhost:9090] (webpack)-dev-server/client?http://localhost:9090 4.29 KiB {main} [built]
[webpack-dev-server] [../../node_modules/webpack-dev-server/client/overlay.js] (webpack)-dev-server/client/overlay.js 3.52 KiB {main} [built]
[webpack-dev-server] [../../node_modules/webpack-dev-server/client/socket.js] (webpack)-dev-server/client/socket.js 1.53 KiB {main} [built]
[webpack-dev-server] [../../node_modules/webpack-dev-server/client/utils/createSocketUrl.js] (webpack)-dev-server/client/utils/createSocketUrl.js 2.91 KiB {main} [built]
[webpack-dev-server] [../../node_modules/webpack-dev-server/client/utils/log.js] (webpack)-dev-server/client/utils/log.js 964 bytes {main} [built]
[webpack-dev-server] [../../node_modules/webpack-dev-server/client/utils/reloadApp.js] (webpack)-dev-server/client/utils/reloadApp.js 1.59 KiB {main} [built]
[webpack-dev-server] [../../node_modules/webpack-dev-server/client/utils/sendMessage.js] (webpack)-dev-server/client/utils/sendMessage.js 402 bytes {main} [built]
[webpack-dev-server] [../../node_modules/webpack-dev-server/node_modules/strip-ansi/index.js] (webpack)-dev-server/node_modules/strip-ansi/index.js 161 bytes {main} [built]
[webpack-dev-server] [../../node_modules/webpack/hot sync ^\.\/log$] (webpack)/hot sync nonrecursive ^\.\/log$ 170 bytes {main} [built]
[webpack-dev-server] [../../node_modules/webpack/hot/dev-server.js] (webpack)/hot/dev-server.js 1.59 KiB {main} [built]
[webpack-dev-server] [../../node_modules/webpack/hot/emitter.js] (webpack)/hot/emitter.js 75 bytes {main} [built]
[webpack-dev-server] [./dev/DevApp.tsx] 484 bytes {main} [built]
[webpack-dev-server]     + 154 hidden modules
[webpack-dev-server] Child HtmlWebpackCompiler:
[webpack-dev-server]      1 asset
[webpack-dev-server]     Entrypoint HtmlWebpackPlugin_0 = __child-HtmlWebpackPlugin_0
[webpack-dev-server]     [../../node_modules/html-webpack-plugin/lib/loader.js!./dev/index.html] /Users/ak/-/projects/hash/github/blockprotocol/node_modules/html-webpack-plugin/lib/loader.js!./dev/index.html 455 bytes {HtmlWebpackPlugin_0} [built]
[webpack-dev-server] ℹ ｢wdm｣: Compiled successfully.
[webpack] yarn build:dev -- --watch --verbose exited with code 0
</code></pre>
</details>

<details>
<summary>After (webpack 5)</summary>
<pre><code>yarn run v1.22.17
$ concurrently -n webpack,webpack-dev-server -c green,cyan "yarn build:dev --watch" "yarn run-dev-server"
$ cross-env NODE_ENV=development webpack-dev-server --config dev/webpack.config.js
$ clean; cross-env NODE_ENV=development webpack --mode development --watch
[webpack] /bin/sh: clean: command not found
[webpack] asset main.js 662 bytes [emitted] (name: main)
[webpack] 
[webpack] ERROR in main
[webpack] Module not found: Error: Can't resolve './src' in '/Users/ak/-/projects/hash/github/blockprotocol/packages/mock-block-dock'
[webpack] resolve './src' in '/Users/ak/-/projects/hash/github/blockprotocol/packages/mock-block-dock'
[webpack]   using description file: /Users/ak/-/projects/hash/github/blockprotocol/packages/mock-block-dock/package.json (relative path: .)
[webpack]     Field 'browser' doesn't contain a valid alias configuration
[webpack]     using description file: /Users/ak/-/projects/hash/github/blockprotocol/packages/mock-block-dock/package.json (relative path: ./src)
[webpack]       no extension
[webpack]         Field 'browser' doesn't contain a valid alias configuration
[webpack]         /Users/ak/-/projects/hash/github/blockprotocol/packages/mock-block-dock/src is not a file
[webpack]       .js
[webpack]         Field 'browser' doesn't contain a valid alias configuration
[webpack]         /Users/ak/-/projects/hash/github/blockprotocol/packages/mock-block-dock/src.js doesn't exist
[webpack]       .json
[webpack]         Field 'browser' doesn't contain a valid alias configuration
[webpack]         /Users/ak/-/projects/hash/github/blockprotocol/packages/mock-block-dock/src.json doesn't exist
[webpack]       .wasm
[webpack]         Field 'browser' doesn't contain a valid alias configuration
[webpack]         /Users/ak/-/projects/hash/github/blockprotocol/packages/mock-block-dock/src.wasm doesn't exist
[webpack]       as directory
[webpack]         existing directory /Users/ak/-/projects/hash/github/blockprotocol/packages/mock-block-dock/src
[webpack]           using description file: /Users/ak/-/projects/hash/github/blockprotocol/packages/mock-block-dock/package.json (relative path: ./src)
[webpack]             using path: /Users/ak/-/projects/hash/github/blockprotocol/packages/mock-block-dock/src/index
[webpack]               using description file: /Users/ak/-/projects/hash/github/blockprotocol/packages/mock-block-dock/package.json (relative path: ./src/index)
[webpack]                 no extension
[webpack]                   Field 'browser' doesn't contain a valid alias configuration
[webpack]                   /Users/ak/-/projects/hash/github/blockprotocol/packages/mock-block-dock/src/index doesn't exist
[webpack]                 .js
[webpack]                   Field 'browser' doesn't contain a valid alias configuration
[webpack]                   /Users/ak/-/projects/hash/github/blockprotocol/packages/mock-block-dock/src/index.js doesn't exist
[webpack]                 .json
[webpack]                   Field 'browser' doesn't contain a valid alias configuration
[webpack]                   /Users/ak/-/projects/hash/github/blockprotocol/packages/mock-block-dock/src/index.json doesn't exist
[webpack]                 .wasm
[webpack]                   Field 'browser' doesn't contain a valid alias configuration
[webpack]                   /Users/ak/-/projects/hash/github/blockprotocol/packages/mock-block-dock/src/index.wasm doesn't exist
[webpack] 
[webpack] webpack 5.68.0 compiled with 1 error in 29 ms
[webpack-dev-server] &lt;w> [webpack-dev-server] "hot: true" automatically applies HMR plugin, you don't have to add it manually to your webpack configuration.
[webpack-dev-server] &lt;i> [webpack-dev-server] Project is running at:
[webpack-dev-server] &lt;i> [webpack-dev-server] Loopback: http://localhost:9090/
[webpack-dev-server] &lt;i> [webpack-dev-server] On Your Network (IPv4): http://192.168.15.179:9090/
[webpack-dev-server] &lt;i> [webpack-dev-server] On Your Network (IPv6): http://[fe80::1]:9090/
[webpack-dev-server] &lt;i> [webpack-dev-server] Content not from webpack is served from '/Users/ak/-/projects/hash/github/blockprotocol/packages/mock-block-dock/dev' directory
[webpack-dev-server] &lt;i> [webpack-dev-middleware] wait until bundle finished: /
[webpack-dev-server] asset main.js 3.51 MiB [emitted] (name: main)
[webpack-dev-server] asset index.html 231 bytes [emitted]
[webpack-dev-server] runtime modules 27 KiB 12 modules
[webpack-dev-server] orphan modules 15.9 KiB [orphan] 10 modules
[webpack-dev-server] modules by path ../../node_modules/core-js/ 86.3 KiB 120 modules
[webpack-dev-server] modules by path ../../node_modules/@babel/runtime/ 4.33 KiB 14 modules
[webpack-dev-server] modules by path ../../node_modules/webpack-dev-server/client/ 56.8 KiB 12 modules
[webpack-dev-server] modules by path ./ 23.8 KiB 11 modules
[webpack-dev-server] modules by path ../../node_modules/uuid/dist/esm-browser/*.js 3.24 KiB 5 modules
[webpack-dev-server] modules by path ../../node_modules/webpack/hot/*.js 4.3 KiB 4 modules
[webpack-dev-server] modules by path ../../node_modules/html-entities/lib/*.js 81.3 KiB 4 modules
[webpack-dev-server] modules by path ../../node_modules/scheduler/ 26.3 KiB 4 modules
[webpack-dev-server] modules by path ../../node_modules/react/ 70.6 KiB 2 modules
[webpack-dev-server] modules by path ../../node_modules/react-dom/ 875 KiB 2 modules
[webpack-dev-server] + 4 modules
[webpack-dev-server] webpack 5.68.0 compiled successfully in 949 ms
</code></pre>
</details>

Do we actually use `dev` in that package? Or is it just copy-pasted from `block-template` and `block-template` is the entry point?

---

Related precursor PRs: #165, #169 and #178.

---

`block-template@0.0.6` and `mock-block-dock@0.0.5` will be released on merge.

### CODE REVIEW

File: packages/block-template/webpack-dev-server.config.js [hunk 1]
```diff
@@ -2,7 +2,9 @@ const HtmlWebpackPlugin = require("html-webpack-plugin");
 const webpack = require("webpack");
 const config = require("./webpack.config");
 
+/** @type import("webpack").Configuration */
 module.exports = {
+  devtool: "eval-cheap-module-source-map",
```
Comment (kachkaev) [inline packages/block-template/webpack-dev-server.config.js:7]: Replaces `-d` option` in `package.json` → `scripts`→ `... webpack-dev-server ...`. The new version of the server requires a value after `-d` (or `--devtool`), so leaving the command as is produces an exception.

File: packages/block-template/webpack-dev-server.config.js [hunk 2]
```diff
@@ -13,13 +15,10 @@ module.exports = {
     new webpack.EnvironmentPlugin({
       "process.env.NODE_ENV": process.env.NODE_ENV,
     }),
-    new webpack.NamedModulesPlugin(),
```
Comment (kachkaev) [inline packages/block-template/webpack-dev-server.config.js:16]: Replaced by

```js
optimization: {
  moduleIds: "named",
},
```

File: packages/block-template/webpack-dev-server.config.js [hunk 3]
```diff
@@ -31,7 +30,12 @@ module.exports = {
         "sentry-trace",
       ],
     },
+    hot: true,
     open: process.env.BROWSER !== "none",
+    port: 9090,
+    static: {
+      directory: __dirname,
+    },
```
Comment (kachkaev) [inline packages/block-template/webpack-dev-server.config.js:38]: Replaces `contentBase: __dirname`

File: packages/block-template/webpack-main.config.js [hunk 4]
```diff
@@ -7,7 +7,7 @@
 const webpack = require("webpack");
 const WebpackAssetsManifest = require("webpack-assets-manifest");
 const { BundleAnalyzerPlugin } = require("webpack-bundle-analyzer");
-const CopyWebpackPlugin = require("copy-webpack-plugin");
+const CopyPlugin = require("copy-webpack-plugin");
```
Comment (kachkaev) [inline packages/block-template/webpack-main.config.js:10]: Aligned with [docs](https://webpack.js.org/plugins/copy-webpack-plugin/)

File: packages/block-template/webpack-main.config.js [hunk 5]
```diff
@@ -20,9 +20,11 @@ module.exports = {
       openAnalyzer: false,
       reportFilename: "webpack-bundle-analyzer-report.html",
     }),
-    new WebpackAssetsManifest(),
+    new WebpackAssetsManifest({
+      output: "manifest.json",
```
Comment (kachkaev) [inline packages/block-template/webpack-main.config.js:24]: The default value [was changed](https://github.com/webdeveric/webpack-assets-manifest#new-in-version-5) to `output-manifest.json`. I added a config option for better backwards-compatibility, but I can remove it now or revisit later.

File: packages/block-template/webpack-main.config.js [hunk 6]
```diff
@@ -20,9 +20,11 @@ module.exports = {
...
+    }),
     new StatsPlugin(),
-    new CopyWebpackPlugin([{ from: "./public/", to: "./public/" }]),
+    new CopyPlugin({ patterns: [{ from: "./public/", to: "./public/" }] }),
```
Comment (kachkaev) [inline packages/block-template/webpack-main.config.js:27]: Breaking change in config shape

**Review by CiaranMn — APPROVED** [review]
Thanks Alex, great stuff! :)

re. `dev` command in `mock-block-dock`: this is useful for testing stuff when we add anything new to the package, although really we should be dealing with this through a bunch of tests rather than a manual `dev` command.

One problem is that the `"clean` part at the start of `build:dev` should in fact be `"yarn clean` - this is what is causing the first error. Not sure about the rest, but we don't need to deal with it now, and certainly not in this PR.

**Pull request merged**